### ISSUE TITLE
Option/Maybe Either and Future Monads in JavaScript Python Ruby Swift and Scala

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#8404]
Option/Maybe, Either, and Future Monads in JavaScript, Python, Ruby, Swift, and Scala<br>
<br>
https://ift.tt/x5Qu6bd<br>
<br>
Kevin Bloch<br>
<br>
<div>
<div>
<div>
<div>
<div>
<p>This monad tutorial gives a brief explanation of monads and shows how to implement the most useful ones in five different programming languages&mdash;if you&rsquo;re looking for monads in <a href="https://www.toptal.com/javascript">JavaScript</a>, monads in <a href="https://www.toptal.com/python">Python</a>, monads in <a href="https://www.toptal.com/ruby">Ruby</a>, monads in <a href="https://www.toptal.com/swift">Swift</a>, and/or monads in <a href="https://www.toptal.com/scala">Scala</a>, or to compare any implementations, you&rsquo;re reading the right article!</p>
<p>Using these monads you will get rid of a series of bugs like null-pointer exceptions, unhandled exceptions, and race conditions.</p>
<p>This is what I cover below:</p>
<ul>
<li>An introduction into category theory</li>
<li>The definition of a monad</li>
<li>Implementations of the Option (&ldquo;Maybe&rdquo;) monad, Either monad, and Future monad, plus a sample program leveraging them, in JavaScript, Python, Ruby, Swift, and Scala</li>
</ul>
<p>Let&rsquo;s get started! Our first stop is category theory, which is the basis for monads.</p>
<h2>Introduction to Category Theory</h2>
<p>Category theory is a mathematical field that was actively developed in the middle of the 20th century. Now it is the basis of many functional programming concepts including the monad. Let&rsquo;s take a quick look into some category theory concepts, tuned for software development terminology.</p>
<p>So there are three core concepts that define a <em>category:</em></p>
<ol>
<li>
<em>Type</em> is just as we see it in statically typed languages. Examples: <code>Int</code>, <code>String</code>, <code>Dog</code>, <code>Cat</code>, etc.</li>
<li>
<em>Functions</em> connect two types. Therefore, they can be represented as an arrow from one type to another type, or to themselves. Function $f$ from type $T$ to type $U$ can be denoted as $f: T \to U$. You can think of it as a programming language function that takes an argument of type $T$ and returns a value of type $U$.</li>
<li>
<em>Composition</em> is an operation, denoted by the $\cdot$ operator, that builds new functions from existing ones. In a category, it&rsquo;s always guaranteed for any functions $f: T \to U$ and $g: U \to V$ there exists a unique function $h: T \to V$. This function is denoted as $f \cdot g$. The operation effectively maps a pair of functions to another function. In programming languages, this operation is, of course, always possible. For instance, if you have a function that returns a length of a string&#8202;&mdash;$strlen: String \to Int$&mdash;and a function that tells if the number is even&#8202;&mdash;$even: Int \to Boolean$&mdash;then you can make a function $even{\_}strlen: String \to Boolean$ which tells if the length of the <code>String</code> is even. In this case $even{\_}strlen = even \cdot strlen$. Composition implies two features:
<ol>
<li>Associativity: $f \cdot g \cdot h = (f \cdot g) \cdot h = f \cdot (g \cdot h)$</li>
<li>The existence of an identity function: $\forall T: \exists f: T \to T$, or in plain English, for every type $T$ there exists a function that maps $T$ to itself.</li>
</ol>
</li>
</ol>
<p>So let&rsquo;s take a look at a simple category.</p>
<figure><img alt="A simple category involving String, Int, and Double, and some functions among them."></figure>
<p><em>Side note: We&rsquo;re assuming that <code>Int</code>, <code>String</code> and all other types here are guaranteed to be non-null, i.e., the null value does not exist.</em></p>
<p><em>Side note 2: This is actually only</em> part <em>of a category, but that is all we want for our discussion, because it has all the essential parts we need and the diagram is less cluttered this way. The real category would also have all composed functions like $roundToString: Double \to String = intToString \cdot round$, to satisfy the composition clause of categories.</em></p>
<p>You might notice that functions in this category are super simple. In fact it&rsquo;s almost impossible to have a bug in these functions. There are no nulls, no exceptions, just the arithmetic and working with memory. So the only bad thing that can happen is processor or memory failure&mdash;in which case you need to crash the program anyway&mdash;but that happens very seldomly.</p>
<p>Wouldn&rsquo;t it be nice if all our code just worked at this level of stability? Absolutely! But what about I/O, for example? We definitely can&rsquo;t live without it. Here is where monad solutions come to the rescue: They isolate all unstable operations into super-small and very well-audited pieces of code&mdash;then you can use stable calculations in your whole app!</p>
<h2>Enter Monads</h2>
<p>Let&rsquo;s call unstable behavior like I/O a <em>side effect</em>. Now we want to be able to work with all our previously defined functions like <code>length</code> and types like <code>String</code> in a stable way in the presence of this <em>side effect</em>.</p>
<p>So let&rsquo;s start with an empty category $M[A]$ and make it into a category that will have values with one particular type of side effect and also values with no side effects. Let&rsquo;s assume we&rsquo;ve defined this category and it&rsquo;s empty. Right now there&rsquo;s nothing useful we can do with it, so to make it useful, we&rsquo;ll follow these three steps:</p>
<ol>
<li>Fill it with values of the types from category $A$, like <code>String</code>, <code>Int</code>, <code>Double</code>, etc. (green boxes in the diagram below)</li>
<li>Once we have these values, we still cannot do anything meaningful with them, so we need a way to take each function $f: T \to U$ from $A$ and create a function $g: M[T] \to M[U]$ (blue arrows in the diagram below). Once we have these functions, we can do everything with the values in category $M[A]$ that we were able to do in category $A$.</li>
<li>Now that we have a brand new $M[A]$ category, a new class of functions emerge with signature $h: T \to M[U]$ (red arrows in the diagram below). They emerge as a result of promoting values in step one as part of our codebase, i.e., we write them as needed; these are the main things that will differentiate working with $M[A]$ versus working with $A$. The final step will be to make these functions work well on types in $M[A]$ as well, i.e., being able to derive function $m: M[T] \to M[U]$ from $h: T \to M[U]$</li>
</ol>
<figure><img alt="Creating a new category: Categories A and M[A], plus a red arrow from A's Double to M[A]'s Int, labelled &quot;roundAsync&quot;. M[A] reuses every value and function of A at this point."></figure>
<p>So let&rsquo;s start off by defining two ways of promoting values of $A$ types to values of $M[A]$ types: one function with no side effects and one with side effects.</p>
<ol>
<li>The first is called $pure$ and is defined for each value of a stable category: $pure: T \to M[T]$. The resulting $M[T]$ values won&rsquo;t have any side effects, hence this function is called $pure$. E.g., for an I/O monad, $pure$ will return some value immediately with no possibility of failure.</li>
<li>The second is called $constructor$ and, unlike $pure$, returns $M[T]$ with some side effects. An example of such a $constructor$ for an async I/O monad could be a function that fetches some data from the web and returns it as a <code>String</code>. The value returned by $constructor$ will have type $M[String]$ in this case.</li>
</ol>
<p>Now that we have two ways of promoting values into $M[A]$, it&rsquo;s up to you as a programmer to choose which function to use, depending on your program goals. Let&rsquo;s consider an example here: You want to fetch an HTML page like https://ift.tt/x5Qu6bd and for this you make a function $fetch$. Since anything could go wrong while fetching it&mdash;think network failures, etc.&mdash;you&rsquo;ll use $M[String]$ as the return type of this function. So it will look something like $fetch: String \to M[String]$ and somewhere in the function body there we will use $constructor$ for $M$.</p>
<p>Now let&rsquo;s assume we make a mock function for testing: $fetchMock: String \to M[String]$. It still has the same signature, but this time we just inject the resulting HTML page inside the body of $fetchMock$ without doing any unstable network operations. So in this case we just use $pure$ in the implementation of $fetchMock$.</p>
<p>As the next step, we need a function that safely promotes any arbitrary function $f$ from the category $A$ to $M[A]$ (blue arrows in a diagram). This function is called $map: (T \to U) \to (M[T] \to M[U])$.</p>
<p>Now we have a category (which can have side effects if we use $constructor$), which also has all functions from the stable category, meaning they are stable in $M[A]$ as well. You might notice that we explicitly introduced another class of functions like $f: T \to M[U]$. E.g., $pure$ and $constructor$ are examples of such functions for $U = T$, but there obviously could be more, like if we were to use $pure$ and then $map$. So, generally, we need a way to deal with arbitrary functions in the form $f: T \to M[U]$.</p>
<p>If we want to make a new function based on $f$ that could be applied to $M[T]$, we could try to use $map$. But that will bring us to function $g: M[T] \to M[M[U]]$, which is no good since we don&rsquo;t want to have one more category $M[M[A]]$. To deal with this problem, we introduce one last function: $flatMap: (T \to M[U]) \to (M[T] \to M[U])$.</p>
<p>But why would we want to do that? Let&rsquo;s assume we&rsquo;re after step 2, i.e., we have $pure$, $constructor$, and $map$. Say we want to grab an HTML page from toptal.com, then scan all the URLs there as well as fetch them. I&rsquo;d make a function $fetch: String \to M[String]$ that fetches just one URL and returns an HTML page.</p>
<p>Then I&rsquo;d apply this function to a URL and get a page from toptal.com, which is $x: M[String]$. Now, I do some transformation on $x$ and finally arrive at some URL $u: M[String]$. I want to apply function $fetch$ to it, but I can&rsquo;t, because it takes type $String$, not $M[String]$. That&rsquo;s why we need $flatMap$ to convert $fetch: String \to M[String]$ to $m_fetch: M[String] \to M[String]$.</p>
<p>Now that we have completed all three steps, we can actually compose any value transformations we need. For instance, if you have value $x$ of type $M[T]$ and $f: T \to U$, you can use $map$ to apply $f$ to value $x$ and get value $y$ of type $M[U]$. That way any transformation of values can be done in a 100 percent bug-free way, as long as the $pure$, $constructor$, $map$ and $flatMap$ implementations are bug-free.</p>
<p>So instead of dealing with some nasty effects each time you encounter them in your codebase, you just need to make sure that only these four functions are implemented correctly. At the end of the program, you will get just one $M[X]$ where you can safely unwrap the value $X$ and handle all error cases.</p>
<p>This is what a monad is&#8202;: &#8202;a thing that implements $pure$, $map$, and $flatMap$. (Actually $map$ can be derived from $pure$ and $flatMap$, but it&rsquo;s very useful and widespread function, so I didn&rsquo;t omit it from the definition.)</p>
<h2>The Option Monad, a.k.a. the Maybe Monad</h2>
<p>Okay, let&rsquo;s dive into the practical implementation and usage of monads. The first really helpful monad is the Option monad. If you&rsquo;re coming from classic programming languages, you&rsquo;ve probably encountered many crashes because of the infamous null pointer error. Tony Hoare, the inventor of null, calls this invention &ldquo;The Billion Dollar Mistake&rdquo;:</p>
<blockquote>
<p>This has led to innumerable errors, vulnerabilities, and system crashes, which have probably caused a billion dollars of pain and damage in the last forty years.</p>
</blockquote>
<p>So let&rsquo;s try to improve on that. The Option monad either holds some non-null value, or no value. Pretty similar to a null value, but having this monad, we can safely use our well-defined functions without being afraid of the null pointer exception. Let&rsquo;s take a look at implementations in different languages:</p>
<div>
<h2>JavaScript&mdash;Option Monad/Maybe Monad</h2>
<div>
<pre><code>
class Monad {
  // pure :: a -&gt; M a
  pure = () =&gt; { throw "pure method needs to be implemented" }
  
  // flatMap :: # M a -&gt; (a -&gt; M b) -&gt; M b
  flatMap = (x) =&gt; { throw "flatMap method needs to be implemented" }

  // map :: # M a -&gt; (a -&gt; b) -&gt; M b
  map = f =&gt; this.flatMap(x =&gt; new this.pure(f(x)))
}

export class Option extends Monad {
  // pure :: a -&gt; Option a
  pure = (value) =&gt; {
    if ((value === null) || (value === undefined)) {
      return none;
    }
    return new Some(value)
  }

  // flatMap :: # Option a -&gt; (a -&gt; Option b) -&gt; Option b
  flatMap = f =&gt; 
    this.constructor.name === 'None' ? 
    none : 
    f(this.value)

  // equals :: # M a -&gt; M a -&gt; boolean
  equals = (x) =&gt; this.toString() === x.toString()
}

class None extends Option {
  toString() {
    return 'None';
  }
}
// Cached None class value
export const none = new None()
Option.pure = none.pure

export class Some extends Option {
  constructor(value) {
    super();
    this.value = value;
  }

  toString() {
    return `Some(${this.value})`
  }
}
</code></pre>
</div>
<h2>Python&mdash;Option Monad/Maybe Monad</h2>
<div>
<pre><code>
class Monad:
  # pure :: a -&gt; M a
  @staticmethod
  def pure(x):
    raise Exception("pure method needs to be implemented")
  
  # flat_map :: # M a -&gt; (a -&gt; M b) -&gt; M b
  def flat_map(self, f):
    raise Exception("flat_map method needs to be implemented")

  # map :: # M a -&gt; (a -&gt; b) -&gt; M b
  def map(self, f):
    return self.flat_map(lambda x: self.pure(f(x)))

class Option(Monad):
  # pure :: a -&gt; Option a
  @staticmethod
  def pure(x):
    return Some(x)

  # flat_map :: # Option a -&gt; (a -&gt; Option b) -&gt; Option b
  def flat_map(self, f):
    if self.defined:
      return f(self.value)
    else:
      return nil

class Some(Option):
  def __init__(self, value):
    self.value = value
    self.defined = True

class Nil(Option):
  def __init__(self):
    self.value = None
    self.defined = False

nil = Nil()
</code></pre>
</div>
<h2>Ruby&mdash;Option Monad/Maybe Monad</h2>
<div>
<pre><code>
class Monad
  # pure :: a -&gt; M a
  def self.pure(x)
    raise StandardError("pure method needs to be implemented")
  end
  
  # pure :: a -&gt; M a
  def pure(x)
    self.class.pure(x)
  end
    
  def flat_map(f)
    raise StandardError("flat_map method needs to be implemented")
  end

  # map :: # M a -&gt; (a -&gt; b) -&gt; M b
  def map(f)
    flat_map(-&gt; (x) { pure(f.call(x)) })
  end
end

class Option &lt; Monad

  attr_accessor :defined, :value

  # pure :: a -&gt; Option a
  def self.pure(x)
    Some.new(x)
  end

  # pure :: a -&gt; Option a
  def pure(x)
    Some.new(x)
  end
  
  # flat_map :: # Option a -&gt; (a -&gt; Option b) -&gt; Option b
  def flat_map(f)
    if defined
      f.call(value)
    else
      $none
    end
  end
end

class Some &lt; Option
  def initialize(value)
    @value = value
    @defined = true
  end
end

class None &lt; Option
  def initialize()
    @defined = false
  end
end

$none = None.new()
</code></pre>
</div>
<h2>Swift&mdash;Option Monad/Maybe Monad</h2>
<div>
<pre><code>
import Foundation

enum Maybe&lt;A&gt; {
  case None
  case Some(A)
  
  static func pure&lt;B&gt;(_ value: B) -&gt; Maybe&lt;B&gt; {
    return .Some(value)
  }
  
  func flatMap&lt;B&gt;(_ f: (A) -&gt; Maybe&lt;B&gt;) -&gt; Maybe&lt;B&gt; {
    switch self {
    case .None:
      return .None
    case .Some(let value):
      return f(value)
    }
  }
  
  func map&lt;B&gt;(f: (A) -&gt; B) -&gt; Maybe&lt;B&gt; {
    return self.flatMap { type(of: self).pure(f($0)) }
  }
}
</code></pre>
</div>
<h2>Scala&mdash;Option Monad/Maybe Monad</h2>
<div>
<pre><code>
import language.higherKinds

trait Monad[M[_]] {
  def pure[A](a: A): M[A]
  def flatMap[A, B](ma: M[A])(f: A =&gt; M[B]): M[B]
  
  def map[A, B](ma: M[A])(f: A =&gt; B): M[B] =
    flatMap(ma)(x =&gt; pure(f(x)))
}
object Monad {
def apply[F[_]](implicit M: Monad[F]): Monad[F] = M
  
  implicit val myOptionMonad = new Monad[MyOption] {
    def pure[A](a: A) = MySome(a)
    def flatMap[A, B](ma: MyOption[A])(f: A =&gt; MyOption[B]): MyOption[B] = ma match {
      case MyNone =&gt; MyNone
      case MySome(a) =&gt; f(a)
    }
  }
}

sealed trait MyOption[+A] {
  def flatMap[B](f: A =&gt; MyOption[B]): MyOption[B] =
    Monad[MyOption].flatMap(this)(f)

  def map[B](f: A =&gt; B): MyOption[B] =
    Monad[MyOption].map(this)(f)
}

case object MyNone extends MyOption[Nothing]
case class MySome[A](x: A) extends MyOption[A]
</code></pre>
</div>
</div>
<p>We start off by implementing a <code>Monad</code> class that will be the base for all of our monad implementations. Having this class is very handy, because by implementing just two of its methods&#8202;&mdash;<code>pure</code> and <code>flatMap</code>&mdash;for a specific monad, you will get many methods for free (we limit them to simply the <code>map</code> method in our examples, but generally there are many other useful methods, like <code>sequence</code> and <code>traverse</code> for working with arrays of <code>Monad</code>s).</p>
<p>We can express <code>map</code> as composition of <code>pure</code> and <code>flatMap</code>. You can see from <code>flatMap</code>&rsquo;s signature $flatMap: (T \to M[U]) \to (M[T] \to M[U])$ that it&rsquo;s really close to $map: (T \to U) \to (M[T] \to M[U])$. The difference is the additional $M$ in the middle, but we can use the <code>pure</code> function to convert $U$ into $M[U]$. That way we express <code>map</code> in terms of <code>flatMap</code> and <code>pure</code>.</p>
<p>This works well for Scala, because it has an advanced type system. It also works well for JS, Python, and Ruby, because they are dynamically typed. Unfortunately, it doesn&rsquo;t work for Swift, because it&rsquo;s statically typed and doesn&rsquo;t have advanced type features like <a href="https://yz.mit.edu/wp/true-scala-complexity/">higher-kinded types</a>, so for Swift we&rsquo;ll have to implement <code>map</code> for each monad.</p>
<p>Also note that the Option monad is already a <em>de facto</em> standard for languages like Swift and Scala, so we use slightly different names for our monad implementations.</p>
<p>Now that we have a base <code>Monad</code> class, let&rsquo;s get to our Option monad implementations. As previously mentioned, the basic idea is that Option either holds some value (called <code>Some</code>) or or doesn&rsquo;t hold any value at all (<code>None</code>).</p>
<p>The <code>pure</code> method simply promotes a value to <code>Some</code>, while the <code>flatMap</code> method checks the current value of the <code>Option</code>&#8202;&mdash;&#8202;if it&rsquo;s <code>None</code> then it returns <code>None</code>, and if it&rsquo;s <code>Some</code> with an underlying value&#8202;, &#8202;it extracts the underlying value, applies <code>f()</code> to it and returns a result.</p>
<p>Note that just using these two functions and <code>map</code>, it&rsquo;s impossible to get into a null pointer exception&mdash;ever. (The problem <em>could</em> potentially arise in our implementation of the <code>flatMap</code> method, but that&rsquo;s just a couple of lines in our code that we check once. After that, we just use our Option monad implementation throughout our code in thousands of places and don&rsquo;t have to fear the null pointer exception at all.)</p>
<h2>The Either Monad</h2>
<p>Let&rsquo;s dive into the second monad&#8202;: &#8202;Either. This is basically the same as the Option monad, but with <code>Some</code> called <code>Right</code> and <code>None</code> called <code>Left</code>. But this time, <code>Left</code> is also allowed to have an underlying value.</p>
<p>We need that because it&rsquo;s very convenient to express throwing an exception. If an exception occurred, then the value of the <code>Either</code> will be <code>Left(Exception)</code>. The <code>flatMap</code> function doesn&rsquo;t progress if the value is <code>Left</code>, which repeats the semantics of throwing exceptions: If an exception happened, we stop further execution.</p>
<div>
<h2>JavaScript&mdash;Either Monad</h2>
<div>
<pre><code>
import Monad from './monad';

export class Either extends Monad {
  // pure :: a -&gt; Either a
  pure = (value) =&gt; {
    return new Right(value)
  }

  // flatMap :: # Either a -&gt; (a -&gt; Either b) -&gt; Either b
  flatMap = f =&gt; 
    this.isLeft() ? 
    this : 
    f(this.value)

  isLeft = () =&gt; this.constructor.name === 'Left'
}

export class Left extends Either {
  constructor(value) {
    super();
    this.value = value;
  }

  toString() {
    return `Left(${this.value})`
  }
}

export class Right extends Either {
  constructor(value) {
    super();
    this.value = value;
  }

  toString() {
    return `Right(${this.value})`
  }
}

// attempt :: (() -&gt; a) -&gt; M a 
Either.attempt = f =&gt; {
    try {
      return new Right(f())
    } catch(e) {
      return new Left(e)
    }
}
Either.pure = (new Left(null)).pure
</code></pre>
</div>
<h2>Python&mdash;Either Monad</h2>
<div>
<pre><code>
from monad import Monad

class Either(Monad):
  # pure :: a -&gt; Either a
  @staticmethod
  def pure(value):
    return Right(value)

  # flat_map :: # Either a -&gt; (a -&gt; Either b) -&gt; Either b
  def flat_map(self, f):
    if self.is_left:
      return self
    else:
      return f(self.value)

class Left(Either):
  def __init__(self, value):
    self.value = value
    self.is_left = True

class Right(Either):
  def __init__(self, value):
    self.value = value
self.is_left = False
</code></pre>
</div>
<h2>Ruby&mdash;Either Monad</h2>
<div>
<pre><code>
require_relative './monad'

class Either &lt; Monad

  attr_accessor :is_left, :value

  # pure :: a -&gt; Either a
  def self.pure(value)
    Right.new(value)
  end

  # pure :: a -&gt; Either a
  def pure(value)
    self.class.pure(value)
  end

  # flat_map :: # Either a -&gt; (a -&gt; Either b) -&gt; Either b
  def flat_map(f)
    if is_left
      self
    else
      f.call(value)
    end
  end
end

class Left &lt; Either
  def initialize(value)
    @value = value
    @is_left = true
  end
end

class Right &lt; Either
  def initialize(value)
    @value = value
    @is_left = false
  end
end
</code></pre>
</div>
<h2>Swift&mdash;Either Monad</h2>
<div>
<pre><code>
import Foundation

enum Either&lt;A, B&gt; {
  case Left(A)
  case Right(B)
  static func pure&lt;C&gt;(_ value: C) -&gt; Either&lt;A, C&gt; {
    return Either&lt;A, C&gt;.Right(value)
  }
  
  func flatMap&lt;D&gt;(_ f: (B) -&gt; Either&lt;A, D&gt;) -&gt; Either&lt;A, D&gt; {
    switch self {
    case .Left(let x):
      return Either&lt;A, D&gt;.Left(x)
    case .Right(let x):
      return f(x)
    }
  }
  
  func map&lt;C&gt;(f: (B) -&gt; C) -&gt; Either&lt;A, C&gt; {
    return self.flatMap { Either&lt;A, C&gt;.pure(f($0)) }
  }
}
</code></pre>
</div>
<h2>Scala&mdash;Either Monad</h2>
<div>
<pre><code>
package monad

sealed trait MyEither[+E, +A] {
  def flatMap[EE &gt;: E, B](f: A =&gt; MyEither[EE, B]): MyEither[EE, B] =
    Monad[MyEither[EE, ?]].flatMap(this)(f)

  def map[EE &gt;: E, B](f: A =&gt; B): MyEither[EE, B] =
    Monad[MyEither[EE, ?]].map(this)(f)
  
}
case class MyLeft[E](e: E) extends MyEither[E, Nothing]
case class MyRight[A](a: A) extends MyEither[Nothing, A]

// ...

implicit def myEitherMonad[E] = new Monad[MyEither[E, ?]] {
  def pure[A](a: A) = MyRight(a)

  def flatMap[A, B](ma: MyEither[E, A])(f: A =&gt; MyEither[E, B]): MyEither[E, B] = ma match {
    case MyLeft(a) =&gt; MyLeft(a)
    case MyRight(b) =&gt; f(b)
  }
}
</code></pre>
</div>
</div>
<p>Also note that it&rsquo;s easy to catch exceptions: All you have to do is map <code>Left</code> to <code>Right</code>. (Although, we don&rsquo;t do it in our examples, for brevity.)</p>
<h2>The Future Monad</h2>
<p>Let&rsquo;s explore the last monad that we need: the &#8202;Future monad. The Future monad is basically a container for a value which is either available now or will be available in the near future. You can make chains of Futures with <code>map</code> and <code>flatMap</code> that will wait for the Future value to be resolved before executing the next piece of code that depends on the value being resolved first. This is very similar to concept of Promises in JS.</p>
<p>Our design goal now is to bridge the existing async APIs in different languages to one consistent base. It turns out that the easiest design approach is using callbacks in $constructor$.</p>
<p>While the callback design introduced the callback hell problem in JavaScript and other languages, it will not be a problem for us, since we use monads. In fact, the <code>Promise</code> object&mdash;the basis for <a href="https://www.toptal.com/javascript/asynchronous-javascript-async-await-tutorial">JavaScript&rsquo;s solution to callback hell</a>&mdash;is a monad itself!</p>
<p>What about the constructor of the Future monad? Is has this signature:</p>
<p><code>constructor :: ((Either err a -&gt; void) -&gt; void) -&gt; Future (Either err a)</code></p>
<p>Let&rsquo;s split it into pieces. First, let&rsquo;s define:</p>
<p><code>type Callback = Either err a -&gt; void</code></p>
<p>So <code>Callback</code> is a function which takes either an error or a resolved value as an argument, and returns nothing. Now our signature looks like this:</p>
<p><code>constructor :: (Callback -&gt; void) -&gt; Future (Either err a)</code></p>
<p>So we need to supply it a function that returns nothing and triggers a callback as soon as async computation is resolved to either an error or some value. Looks easy enough to make a bridge for any language.</p>
<p>As for the design of the Future monad itself, let&rsquo;s look at its internal structure. The key idea is to have a cache variable that holds a value if the Future monad is resolved, or holds nothing otherwise. You can subscribe to the Future with a callback which will be immediately triggered if value is resolved, or if not, will put the callback into the subscribers list.</p>
<p>Once the Future is resolved, each callback in this list will be triggered exactly once with the resolved value in a separate thread (or as the next function to be executed in the event loop, in the case of JS.) Note that it&rsquo;s crucial to carefully use synchronization primitives, otherwise race conditions are possible.</p>
<p>The basic flow is: You start the async computation supplied as the constructor argument, and point its callback to our internal callback method. In the meantime, you can subscribe to the Future monad and put your callbacks into the queue. Once the computation is done, the internal callback method calls all callbacks in the queue. If you&rsquo;re familiar with Reactive Extensions (RxJS, RxSwift, etc.), they use a very similar approach to async handling.</p>
<p>The public API of the Future monad consists of <code>pure</code>, <code>map</code>, and <code>flatMap</code>, just like in the previous monads. We&rsquo;ll also need a couple of handy methods:</p>
<ol>
<li>
<code>&#8202;async</code>, which takes a synchronous blocking function and executes it on a separate thread, and</li>
<li>
<code>traverse</code>, which takes an array of values and function that maps a value to a <code>Future</code>, and returns a <code>Future</code> of an array of resolved values</li>
</ol>
<p>Let&rsquo;s see how that plays out:</p>
<div>
<h2>JavaScript&mdash;Future Monad</h2>
<div>
<pre><code>
import Monad from './monad';
import { Either, Left, Right } from './either';
import { none, Some } from './option';

export class Future extends Monad {
  // constructor :: ((Either err a -&gt; void) -&gt; void) -&gt; Future (Either err a)
  constructor(f) {
    super();
    this.subscribers = [];
    this.cache = none;
    f(this.callback)
  }

  // callback :: Either err a -&gt; void
  callback = (value) =&gt; {
    this.cache = new Some(value)
    while (this.subscribers.length) {
      const subscriber = this.subscribers.shift();
      subscriber(value)
    }
  }

  // subscribe :: (Either err a -&gt; void) -&gt; void
  subscribe = (subscriber) =&gt; 
    (this.cache === none ? this.subscribers.push(subscriber) : subscriber(this.cache.value))

  toPromise = () =&gt; new Promise(
    (resolve, reject) =&gt;
      this.subscribe(val =&gt; val.isLeft() ? reject(val.value) : resolve(val.value))
  )

  // pure :: a -&gt; Future a
  pure = Future.pure

  // flatMap :: (a -&gt; Future b) -&gt; Future b
  flatMap = f =&gt;
    new Future(
      cb =&gt; this.subscribe(value =&gt; value.isLeft() ? cb(value) : f(value.value).subscribe(cb))
    )
}

Future.async = (nodeFunction, ...args) =&gt; {
  return new Future(cb =&gt; 
    nodeFunction(...args, (err, data) =&gt; err ? cb(new Left(err)) : cb(new Right(data)))
  );
}

Future.pure = value =&gt; new Future(cb =&gt; cb(Either.pure(value)))

// traverse :: [a] -&gt; (a -&gt; Future b) -&gt; Future [b]
Future.traverse = list =&gt; f =&gt;
  list.reduce(
    (acc, elem) =&gt; acc.flatMap(values =&gt; f(elem).map(value =&gt; [...values, value])), 
    Future.pure([])
  )
</code></pre>
</div>
<h2>Python&mdash;Future Monad</h2>
<div>
<pre><code>
from monad import Monad
from option import nil, Some
from either import Either, Left, Right
from functools import reduce
import threading

class Future(Monad):
  # __init__ :: ((Either err a -&gt; void) -&gt; void) -&gt; Future (Either err a)
  def __init__(self, f):
    self.subscribers = []
    self.cache = nil
    self.semaphore = threading.BoundedSemaphore(1)
    f(self.callback)

  # pure :: a -&gt; Future a
  @staticmethod
  def pure(value):
    return Future(lambda cb: cb(Either.pure(value)))

  def exec(f, cb):
    try:
      data = f()
      cb(Right(data))
    except Exception as err:
      cb(Left(err))

  def exec_on_thread(f, cb):
    t = threading.Thread(target=Future.exec, args=[f, cb])
    t.start()

  def async(f):
    return Future(lambda cb: Future.exec_on_thread(f, cb))

  # flat_map :: (a -&gt; Future b) -&gt; Future b
  def flat_map(self, f):
    return Future(
      lambda cb: self.subscribe(
        lambda value: cb(value) if (value.is_left) else f(value.value).subscribe(cb)
      )
    )

  # traverse :: [a] -&gt; (a -&gt; Future b) -&gt; Future [b]
  def traverse(arr):
    return lambda f: reduce(
      lambda acc, elem: acc.flat_map(
        lambda values: f(elem).map(
          lambda value: values + [value]
        )
      ), arr, Future.pure([]))

  # callback :: Either err a -&gt; void
  def callback(self, value):
    self.semaphore.acquire()
    self.cache = Some(value)
    while (len(self.subscribers) &gt; 0):
      sub = self.subscribers.pop(0)
      t = threading.Thread(target=sub, args=[value])
      t.start()
    self.semaphore.release()
  
  # subscribe :: (Either err a -&gt; void) -&gt; void
  def subscribe(self, subscriber):
    self.semaphore.acquire()
    if (self.cache.defined):
      self.semaphore.release()
      subscriber(self.cache.value)
    else:
      self.subscribers.append(subscriber)
self.semaphore.release()
</code></pre>
</div>
<h2>Ruby&mdash;Future Monad</h2>
<div>
<pre><code>
require_relative './monad'
require_relative './either'
require_relative './option'

class Future &lt; Monad
  attr_accessor :subscribers, :cache, :semaphore

  # initialize :: ((Either err a -&gt; void) -&gt; void) -&gt; Future (Either err a)
  def initialize(f)
    @subscribers = []
    @cache = $none
    @semaphore = Queue.new
    @semaphore.push(nil)
    f.call(method(:callback))
  end

  # pure :: a -&gt; Future a
  def self.pure(value)
    Future.new(-&gt; (cb) { cb.call(Either.pure(value)) })
  end

  def self.async(f, *args)
    Future.new(-&gt; (cb) {
      Thread.new {
        begin
          cb.call(Right.new(f.call(*args)))
        rescue =&gt; e
          cb.call(Left.new(e))
        end  
      }
    })
  end

  # pure :: a -&gt; Future a
  def pure(value)
    self.class.pure(value)
  end

  # flat_map :: (a -&gt; Future b) -&gt; Future b
  def flat_map(f)
    Future.new(-&gt; (cb) { 
      subscribe(-&gt; (value) {
        if (value.is_left)
          cb.call(value) 
        else
          f.call(value.value).subscribe(cb)
        end
      }) 
    })
  end

  # traverse :: [a] -&gt; (a -&gt; Future b) -&gt; Future [b]
  def self.traverse(arr, f)
    arr.reduce(Future.pure([])) do |acc, elem|
      acc.flat_map(-&gt; (values) {
        f.call(elem).map(-&gt; (value) { values + [value] })
      })
    end
  end

  # callback :: Either err a -&gt; void
  def callback(value)
    semaphore.pop
    self.cache = Some.new(value)
    while (subscribers.count &gt; 0)
      sub = self.subscribers.shift
      Thread.new {
        sub.call(value)
      }
    end
    semaphore.push(nil)
  end
  
  # subscribe :: (Either err a -&gt; void) -&gt; void
  def subscribe(subscriber)
    semaphore.pop
    if (self.cache.defined)
      semaphore.push(nil)
      subscriber.call(cache.value)
    else
      self.subscribers.push(subscriber)
      semaphore.push(nil)
    end
  end
end
</code></pre>
</div>
<h2>Swift&mdash;Future Monad</h2>
<div>
<pre><code>
import Foundation

let background = DispatchQueue(label: "background", attributes: .concurrent)

class Future&lt;Err, A&gt; {
  typealias Callback = (Either&lt;Err, A&gt;) -&gt; Void
  
  var subscribers: Array&lt;Callback&gt; = Array&lt;Callback&gt;()
  var cache: Maybe&lt;Either&lt;Err, A&gt;&gt; = .None
  var semaphore = DispatchSemaphore(value: 1)

  lazy var callback: Callback = { value in
    self.semaphore.wait()
    self.cache = .Some(value)
    while (self.subscribers.count &gt; 0) {
      let subscriber = self.subscribers.popLast()
      background.async {
        subscriber?(value)
      }
    }
    self.semaphore.signal()
  }
  
  init(_ f: @escaping (@escaping Callback) -&gt; Void) {
    f(self.callback)
  }
  
  func subscribe(_ cb: @escaping Callback) {
    self.semaphore.wait()
    switch cache {
    case .None:
      subscribers.append(cb)
      self.semaphore.signal()
    case .Some(let value):
      self.semaphore.signal()
      cb(value)
    }
  }
  
  static func pure&lt;B&gt;(_ value: B) -&gt; Future&lt;Err, B&gt; {
    return Future&lt;Err, B&gt; { $0(Either&lt;Err, B&gt;.pure(value)) }
  }
  
  func flatMap&lt;B&gt;(_ f: @escaping (A) -&gt; Future&lt;Err, B&gt;) -&gt; Future&lt;Err, B&gt; {
    return Future&lt;Err, B&gt; { [weak self] cb in
      guard let this = self else { return }
      this.subscribe { value in
        switch value {
        case .Left(let err):
          cb(Either&lt;Err, B&gt;.Left(err))
        case .Right(let x):
          f(x).subscribe(cb)
        }
      }
    }
  }
  
  func map&lt;B&gt;(_ f: @escaping (A) -&gt; B) -&gt; Future&lt;Err, B&gt; {
    return self.flatMap { Future&lt;Err, B&gt;.pure(f($0)) }
  }
  
  static func traverse&lt;B&gt;(_ list: Array&lt;A&gt;, _ f: @escaping (A) -&gt; Future&lt;Err, B&gt;) -&gt; Future&lt;Err, Array&lt;B&gt;&gt; {
    return list.reduce(Future&lt;Err, Array&lt;B&gt;&gt;.pure(Array&lt;B&gt;())) { (acc: Future&lt;Err, Array&lt;B&gt;&gt;, elem: A) in
      return acc.flatMap { elems in
        return f(elem).map { val in
          return elems + [val]
        }
      }
    }
  }
}
</code></pre>
</div>
<h2>Scala&mdash;Future Monad</h2>
<div>
<pre><code>
package monad

import java.util.concurrent.Semaphore

class MyFuture[A] {
  private var subscribers: List[MyEither[Exception, A] =&gt; Unit] = List()
  private var cache: MyOption[MyEither[Exception, A]] = MyNone
  private val semaphore = new Semaphore(1)

  def this(f: (MyEither[Exception, A] =&gt; Unit) =&gt; Unit) {
    this()
    f(this.callback _)
  }

  def flatMap[B](f: A =&gt; MyFuture[B]): MyFuture[B] =
    Monad[MyFuture].flatMap(this)(f)

  def map[B](f: A =&gt; B): MyFuture[B] =
    Monad[MyFuture].map(this)(f)

  def callback(value: MyEither[Exception, A]): Unit = {
    semaphore.acquire
    cache = MySome(value)
    subscribers.foreach { sub =&gt;
      val t = new Thread(
        new Runnable {
          def run: Unit = {
            sub(value)
          }
        }
      )
      t.start
    }
    subscribers = List()
    semaphore.release
  }

  def subscribe(sub: MyEither[Exception, A] =&gt; Unit): Unit = {
    semaphore.acquire
    cache match {
      case MyNone =&gt;
        subscribers = sub :: subscribers
        semaphore.release
      case MySome(value) =&gt;
        semaphore.release
        sub(value)
    }
  }
}

object MyFuture {
  def async[B, C](f: B =&gt; C, arg: B): MyFuture[C] =
    new MyFuture[C]({ cb =&gt;
      val t = new Thread(
        new Runnable {
          def run: Unit = {
            try {
              cb(MyRight(f(arg)))
            } catch {
              case e: Exception =&gt; cb(MyLeft(e))
            }
          }
        }
      )
      t.start
    })

  def traverse[A, B](list: List[A])(f: A =&gt; MyFuture[B]): MyFuture[List[B]] = {
    list.foldRight(Monad[MyFuture].pure(List[B]())) { (elem, acc) =&gt;
      Monad[MyFuture].flatMap(acc) ({ values =&gt;
        Monad[MyFuture].map(f(elem)) { value =&gt; value :: values }
      })
    }
  }
}

// ...

implicit val myFutureMonad = new Monad[MyFuture] {
  def pure[A](a: A): MyFuture[A] = 
    new MyFuture[A]({ cb =&gt; cb(myEitherMonad[Exception].pure(a)) })

  def flatMap[A, B](ma: MyFuture[A])(f: A =&gt; MyFuture[B]): MyFuture[B] =
    new MyFuture[B]({ cb =&gt;
      ma.subscribe(_ match {
        case MyLeft(e) =&gt; cb(MyLeft(e))
        case MyRight(a) =&gt; f(a).subscribe(cb)
      })
    })
}
</code></pre>
</div>
</div>
<p>Now, notice how the public API of <code>Future</code> doesn&rsquo;t contain any low-level details like threads, semaphores, or any of that stuff. All you need is basically to supply something with a callback, and that&rsquo;s it!</p>
<h2>Composing a Program from Monads</h2>
<p>Okay, so let&rsquo;s try to use our monads to make an actual program. Say we have a file with a list of URLs and we want to fetch each of these URLs in parallel. Then, we want to cut the responses to 200 bytes each for brevity and print out the result.</p>
<p>We start off by converting existing language APIs to monadic interfaces (see the functions <code>readFile</code> and <code>fetch</code>). Now that we have that, &#8202;we can just compose them to get the final result as one chain. Note that the chain itself is super safe, as all the gory details are contained in monads.</p>
<div>
<h2>JavaScript&mdash;Sample Monad Program</h2>
<div>
<pre><code>
import { Future } from './future';
import { Either, Left, Right } from './either';
import { readFile } from 'fs';
import https from 'https';

const getResponse = url =&gt; 
  new Future(cb =&gt; https.get(url, res =&gt; {
    var body = '';
    res.on('data', data =&gt; body += data);
    res.on('end', data =&gt; cb(new Right(body)));
    res.on('error', err =&gt; cb(new Left(err)))
  }))

const getShortResponse = url =&gt; getResponse(url).map(resp =&gt; resp.substring(0, 200))

Future
  .async(readFile, 'resources/urls.txt')
  .map(data =&gt; data.toString().split("\n"))
  .flatMap(urls =&gt; Future.traverse(urls)(getShortResponse))
.map(console.log)
</code></pre>
</div>
<h2>Python&mdash;Sample Monad Program</h2>
<div>
<pre><code>
import http.client
import threading
import time
import os
from future import Future
from either import Either, Left, Right

conn = http.client.HTTPSConnection("en.wikipedia.org")

def read_file_sync(uri):
  base_dir = os.path.dirname(__file__) #&lt;-- absolute dir the script is in
  path = os.path.join(base_dir, uri)
  with open(path) as f:
    return f.read()

def fetch_sync(uri):
  conn.request("GET", uri)
  r = conn.getresponse()
  return r.read().decode("utf-8")[:200]

def read_file(uri):
  return Future.async(lambda: read_file_sync(uri))

def fetch(uri):
  return Future.async(lambda: fetch_sync(uri))

def main(args=None):
  lines = read_file("../resources/urls.txt").map(lambda res: res.splitlines())
  content = lines.flat_map(lambda urls: Future.traverse(urls)(fetch))
  output = content.map(lambda res: print("\n".join(res)))

if __name__ == "__main__":
main()
</code></pre>
</div>
<h2>Ruby&mdash;Sample Monad Program</h2>
<div>
<pre><code>
require './lib/future'
require 'net/http'
require 'uri'

semaphore = Queue.new

def read(uri)
  Future.async(-&gt; () { File.read(uri) })
end

def fetch(url)
  Future.async(-&gt; () {
    uri = URI(url)
    Net::HTTP.get_response(uri).body[0..200]
  })
end

read("resources/urls.txt")
  .map(-&gt; (x) { x.split("\n") })
  .flat_map(-&gt; (urls) {
    Future.traverse(urls, -&gt; (url) { fetch(url) })
  })
  .map(-&gt; (res) { puts res; semaphore.push(true) })

semaphore.pop
</code></pre>
</div>
<h2>Swift&mdash;Sample Monad Program</h2>
<div>
<pre><code>
import Foundation

enum Err: Error {
  case Some(String)
}

func readFile(_ path: String) -&gt; Future&lt;Error, String&gt; {
  return Future&lt;Error, String&gt; { callback in
    background.async {
      let url = URL(fileURLWithPath: path)
      let text = try? String(contentsOf: url)
      if let res = text {
        callback(Either&lt;Error, String&gt;.pure(res))
      } else {
        callback(Either&lt;Error, String&gt;.Left(Err.Some("Error reading urls.txt")))
      }
    }
  }
}

func fetchUrl(_ url: String) -&gt; Future&lt;Error, String&gt; {
  return Future&lt;Error, String&gt; { callback in
    background.async {
      let url = URL(string: url)
      let task = URLSession.shared.dataTask(with: url!) {(data, response, error) in
        if let err = error {
          callback(Either&lt;Error, String&gt;.Left(err))
          return
        }
        guard let nonEmptyData = data else {
          callback(Either&lt;Error, String&gt;.Left(Err.Some("Empty response")))
          return
        }
        guard let result = String(data: nonEmptyData, encoding: String.Encoding.utf8) else {
          callback(Either&lt;Error, String&gt;.Left(Err.Some("Cannot decode response")))
          return
        }
        let index = result.index(result.startIndex, offsetBy: 200)
        callback(Either&lt;Error, String&gt;.pure(String(result[..&lt;index])))
      }
      task.resume()
    }
  }
}

var result: Any = ""
let _ = readFile("\(projectDir)/Resources/urls.txt")
    .map { data -&gt; [String] in
      data.components(separatedBy: "\n").filter { (line: String) in !line.isEmpty }
    }.flatMap { urls in
        return Future&lt;Error, String&gt;.traverse(urls) { url in
            return fetchUrl(url)
        }
    }.map { responses in
      print(responses)
    }

RunLoop.main.run()
</code></pre>
</div>
<h2>Scala&mdash;Sample Monad Program</h2>
<div>
<pre><code>
import scala.io.Source
import java.util.concurrent.Semaphore
import monad._

object Main extends App {
  val semaphore = new Semaphore(0)

  def readFile(name: String): MyFuture[List[String]] =
    MyFuture.async[String, List[String]](filename =&gt; Source.fromResource(filename).getLines.toList, name)
  
  def fetch(url: String): MyFuture[String] =
    MyFuture.async[String, String](
      uri =&gt; Source.fromURL(uri).mkString.substring(0, 200),
      url
    )
  
  val future = for {
    urls &lt;- readFile("urls.txt")
    entries &lt;- MyFuture.traverse(urls)(fetch _)
  } yield { 
    println(entries)
    semaphore.release
  }

  semaphore.acquire
}
</code></pre>
</div>
</div>
<p>There you have it&mdash;monad solutions in practice. You can find a repo containing all the code from this article <a href="https://github.com/alleycat-at-git/monad">on GitHub</a>.</p>
<h2>Overhead: Done. Benefits: Ongoing</h2>
<p>For this simple monad-based program, it might look like overkill to use all the code that we wrote before. But that&rsquo;s just the initial setup, and it will stay constant in its size. Imagine that from now on, using monads, you can write a lot of async code, not worrying about threads, race conditions, semaphores, exceptions, or null pointers! Awesome!</p>
</div>
</div>
</div>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via Toptal Engineering Blog<br>
<br>
April 25, 2022 at 11:26PM
